---
title: "First steps with NGS data"
author: "Valentin Loux - Olivier Rué"
subtitle: "DUBii - Module 5"
css: css/styles.css
date: "2020/03/09"
output:
    html_document:
      self_contained: true
      number_sections: false
      code_folding: "show"
      toc: true
      toc_depth: 1
      toc_float: true
---



```{r setup, include=FALSE}
options(htmltools.dir.version = FALSE)
library(RefManageR)
BibOptions(check.entries = FALSE, bib.style = "authoryear", style = "markdown",
           dashed = TRUE)
file.name <- "biblio.bib"
bib <- ReadBib(file.name)
```

# Hands-on 1: Preparation of your working directory

## Instruction

- Go in your home directory
- Create a directory called M5 (i.e Module5) and move in
- Create this directory structure:
```{bash, eval=FALSE}
tree ~/M5
[orue@clust-slurm-client M5]$ tree .
.
├── CLEANING
├── FASTQ
├── MAPPING
└── QC

4 directories, 0 files
```



## Correction

```{bash, eval=FALSE}
mkdir -p ~/M5/FASTQ
mkdir -p ~/M5/CLEANING
mkdir -p ~/M5/MAPPING
mkdir -p ~/M5/QC
cd ~/M5
```



# Hands-on 2: Getting raw data 

## Instruction

Get the raw shot read data (Illumina) associated with this article `r Citep(bib, "Allue-Guardiae01052-18")`.

```{r, fig.align="center", out.width="70%", echo=FALSE}
knitr::include_graphics("images/MRA.01052-18.png")
```

- In the "Data availability" section, extract the accession for Illumina data : SRX4909245
- Explore [SRA](https://www.ncbi.nlm.nih.gov/sra/SRX4909245) and [ENA](https://www.ebi.ac.uk/ena/browser/view/SRX4909245)

Get the data by the method of your choice:
- use <code>wget</code> or <code>fasterq-dump</code> from <code>sra-tools</code>

Compress FASTQ files with <code>gzip</code>



## Correction

- Direct download via the web brower
- Using wget :
```{bash, eval=FALSE, class="large"}
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR808/003/SRR8082143/SRR8082143_1.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR808/003/SRR8082143/SRR8082143_2.fastq.gz
```
- Using sra-toolkit
```{bash, eval=FALSE, class="large"}
module load  sra-tools
srun fasterq-dump -S -p SRR8082143 --outdir . --threads 1
```
- enaBrowserTool is also available
```{bash, eval=FALSE, class="large", echo=FALSE}
module load enabrowsertools
srun enaDataGet -f fastq SRR8082143 
```
- Compress FASTQ files
```{bash, eval=FALSE}
gzip *.fastq
```

```{bash, eval=FALSE}
ls -ltrh ~/M5/FASTQ/
total 236M
-rw-rw-r-- 1 orue orue 127M  6 mars  12:32 SRR8082143_2.fastq.gz
-rw-rw-r-- 1 orue orue 109M  6 mars  12:32 SRR8082143_1.fastq.gz

```




# Hands-on 3: Quality control

## Instruction

- Launch FastQC on the paired-end FastQ files of the sample you previously downloaded
- Inspect the results
  - Are the number coherent with the article ?
  - Comment on the quality of the sequencing


## Correction

```{bash, eval=FALSE, class="large"}
cd ~
module load fastqc
srun --cpus-per-task 8 fastqc FASTQ/SRR8082143_1.fastq.gz -o QC/ -t 8
srun --cpus-per-task 8 fastqc FASTQ/SRR8082143_2.fastq.gz -o QC/ -t 8
```

```{bash, eval=FALSE}
ls -ltrh ~/M5/QC
total 1,9M
-rw-rw-r-- 1 orue orue 321K  6 mars  13:23 SRR8082143_1_fastqc.zip
-rw-rw-r-- 1 orue orue 642K  6 mars  13:23 SRR8082143_1_fastqc.html
-rw-rw-r-- 1 orue orue 333K  6 mars  13:23 SRR8082143_2_fastqc.zip
-rw-rw-r-- 1 orue orue 642K  6 mars  13:23 SRR8082143_2_fastqc.html
```


#  Hands-on 4: Reads cleaning with fastp

## Instruction

- Launch fastp on the paired-end FastQ files of the sample you previously downloaded
  - Detect and Remove the classical Illumina adapters
  - Filter reads with :
      - mean quality >= 20 on a sliding window of 4
      -  40% of the bases with a quality >= 15
      - length of the trimmed read >= 100

- Inspect the results
  - How many reads are filtered ?
  - Where do fastp store its reports. Is it configurable ?



## Correction

```{bash, eval=FALSE, class="large"}
module load fastp
cd ~/M5
srun --cpus-per-task 8 fastp -i FASTQ/SRR8082143_1.fastq.gz --in1 FASTQ/SRR8082143_1.fastq.gz --in2 FASTQ/SRR8082143_2.fastq.gz -l 100 --out1 CLEANING/SRR8082143_1.cleaned_filtered.fastq.gz --out2 CLEANING/SRR8082143_2.cleaned_filtered.fastq.gz --unpaired1 CLEANING/SRR8082143_singles.fastq.gz --unpaired2 CLEANING/SRR8082143_singles.fastq.gz -w 1 -j CLEANING/fastp.json -h CLEANING/fastp.html -t 8
```
```{bash, eval=FALSE}
ls -ltrh ~/M5/CLEANING/
total 245M
-rw-rw-r-- 1 orue orue 113M  6 mars  12:59 SRR8082143_1.cleaned_filtered.fastq.gz
-rw-rw-r-- 1 orue orue 162K  6 mars  12:59 fastp.json
-rw-rw-r-- 1 orue orue 525K  6 mars  12:59 fastp.html
-rw-rw-r-- 1 orue orue 2,2M  6 mars  12:59 SRR8082143_singles.fastq.gz
-rw-rw-r-- 1 orue orue 130M  6 mars  12:59 SRR8082143_2.cleaned_filtered.fastq.gz
```


# Hands-on 5: MultiQC

## Instruction

Run MultiQC to obtain a report with fastqc and fastp results

## Correction

```{bash, eval=FALSE}
cd ~/M5
module load multiqc
multiqc -d . -o CLEANING
```

```{bash, eval=FALSE}
ls -ltrh ~/M5/CLEANING/
total 248M
-rw-rw-r-- 1 orue orue 113M  6 mars  12:59 SRR8082143_1.cleaned_filtered.fastq.gz
-rw-rw-r-- 1 orue orue 162K  6 mars  12:59 fastp.json
-rw-rw-r-- 1 orue orue 525K  6 mars  12:59 fastp.html
-rw-rw-r-- 1 orue orue 2,2M  6 mars  12:59 SRR8082143_singles.fastq.gz
-rw-rw-r-- 1 orue orue 130M  6 mars  12:59 SRR8082143_2.cleaned_filtered.fastq.gz
-rw-rw-r-- 1 orue orue 1,2M  6 mars  13:28 multiqc_report.html
drwxrwxr-x 2 orue orue 2,0M  6 mars  13:28 multiqc_data
```



# Hands-on 6: mapping with bwa

## Instruction

- Map the reads to the reference genome <code>/shared/projects/dubii2020/data/module5/seance1/CP031214.1.fasta</code> with <code>bwa</code>

## Correction

```{bash, eval=FALSE}
cd ~/M5
module load bwa
## srun bwa index sequence.fasta
srun --cpus-per-task=33 bwa mem /shared/projects/dubii2020/data/module5/seance1/CP031214.1.fasta CLEANING/SRR8082143_1.cleaned_filtered.fastq.gz CLEANING/SRR8082143_2.cleaned_filtered.fastq.gz -t 32 | samtools view -hbS - > MAPPING/SRR8082143.bam
```

```{bash, eval=FALSE}
ls -ltrh ~/M5/MAPPING/
total 249M
-rw-rw-r-- 1 orue orue 249M  6 mars  13:01 SRR8082143.bam
```

